### PR TITLE
Dedupe pending requests to the same cached chunk

### PR DESCRIPTION
**Pull request by ekzhang**
This reduces needed throughput for large file reads that end up requesting the same ranges of bytes multiple times simultaneously.

Before a request to some `(hash, block_num)` pair is made, it inserts a pending request in the `pending_requests` field of the cache state. This makes it so that only one request to a given chunk is made at a time.

### CODE REVIEW

File: blobnet/src/provider.rs [hunk 1]
```diff
@@ -481,14 +483,46 @@ impl<P> CachedState<P> {
...
+            match rx.recv().await {
+                Ok(result) => return result,
+                Err(_) => return Err(anyhow!("pending request channel dropped").into()),
+            }
+        }
+
+        let result = func().await;
+
+        // Finish all blocked, pending requests.
+        if let Some(pending_tx) = self.pending_requests.lock().remove(&(hash.clone(), n)) {
```
Comment (thundergolfer) [inline blobnet/src/provider.rs:509]: Is there a valid situation where the `.remove` returns `None`? It seems that the entry populator has to be the one to remove the entry.

**Review by thundergolfer — COMMENTED** [review]
Man, you got this out fast! Great stuff

File: blobnet/src/provider.rs (hunk 1 above)
Comment (ekzhang) [reply blobnet/src/provider.rs:509]: Yeah it shouldn't ever return `None`, I'm just being safe :)

Comment (thundergolfer) [reply blobnet/src/provider.rs:509]: Is this a potential use-case for `debug_assert!`?

Comment (ekzhang) [reply blobnet/src/provider.rs:509]: I think it could be!

**Pull request merged**